### ISSUE TITLE
dtrace is not working

### ISSUE THREAD

**Issue by ldldr** [opnsense/core#1726]
I cannot use dtrace on OPNsense 17.1.9/11.0-RELEASE-p10:
```
root@opnsense:~ # dtrace -V
dtrace: Sun D 1.12.1

root@opnsense:~ # dtrace -n 'dtrace:::BEGIN { printf("Hello OPNsense!\n"); }'                                                                          
dtrace: invalid probe specifier dtrace:::BEGIN { printf("Hello OPNsense!\n"); }: "/usr/lib/dtrace/io.d", line 42: operator -> cannot be applied to a forward declaration: no struct devstat definition is available
```
This works on pristine FreeBSD 11.0-RELEASE-p9 and -p11.

I have seen a similar issue here: https://bugs.freenas.org/issues/6521

**Comment by fichtner** [comment]
Could be because of one of the src.conf settings, or the fact that we don't use debug info in the kernel. I have no idea which, but maybe the latter.

https://github.com/opnsense/tools/blob/master/config/17.1/src.conf
https://github.com/opnsense/tools/blob/master/config/17.1/SMP#L6

**Comment by ldldr** [comment]
The necessary module is already loaded:
```
root@opnsense:~ # kldstat | grep dtraceall
48    1 0xffffffff827fc000 33f      dtraceall.ko
```

Citing https://www.freebsd.org/doc/handbook/dtrace-enable.html the following options should be configured to enable dtrace support on amd64:
```
options         KDTRACE_HOOKS
options         DDB_CTF
makeoptions	DEBUG=-g
makeoptions	WITH_CTF=1
options         KDTRACE_FRAME
```
So it appears this cannot be resolved without rebuilding the kernel.

I can understand that you want to remove the debugging info from the kernel for embedded systems (nano). But could you provide a kernel with dtrace support for the other opnsense installation types?

**Comment by fichtner** [comment]
I always wanted to add kernel debug files as an optional set, maybe that will already solve it. Bit this has to wait till August to focus on the 17.7 release.

I can, however, try to provide a test kernel for you within the next few days?

**Comment by ldldr** [comment]
Thanks, dtrace is an extremely valuable tool even for production systems. I appreciate that you keep it on your radar. No need to rush on my account though. I am patient.

**Comment by fichtner** [comment]
Hi @ldldr,

It's time. :)

Please try the following on 17.7 (amd64) to see if dtrace works now:

    # opnsense-code update

(pulls the code for opnsense-update)

    # cd /usr/update && make all install

(installs the code for opnsense-update)

    # opnsense-update -gkfr 17.7

(installs the debug (-g) kernel (-k) forcefully (-f) for version 17.7 (-r))

    # /usr/local/etc/rc.reboot

(reboots to set the kernel as active)

There are more things to do eventually like how we choose a debug kernel over a normal one, when we provide one (currently a one-shot for 17.7), etc.


Cheers,
Franco

**Comment by fichtner** [comment]
The necessary pieces should be in place. We can't speed up our fixed time frame development for 18.1 so closing this issue unless there is something else required.


Cheers,
Franco
